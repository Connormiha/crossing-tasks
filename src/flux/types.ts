export type RIVERSIDE_LEFT_TYPE = 'left';
export type RIVERSIDE_RIGHT_TYPE = 'right';
export type BOAT_TYPE = 'boat';

export const RIVERSIDE_LEFT: RIVERSIDE_LEFT_TYPE = 'left';
export const RIVERSIDE_RIGHT: RIVERSIDE_RIGHT_TYPE = 'right';
export const BOAT: BOAT_TYPE = 'boat';

export interface IMessageState {
  readonly content: string;
  readonly hidden: boolean;
}

export interface IGameState {
  readonly currentGame: string;
  readonly finished: boolean;
  readonly list: string[];
}

export interface ISettingsState {
  readonly volume: number;
}

export interface ICharacterBase<T extends string> {
  id: T;
  name: string;
}

export type ICharactersMap<T> = T extends ICharacterBase<infer S> ? Record<S, T> : never;
//export type ICharactersList<T> = T extends ICharacterBase<infer S> ? (S extends string ? T[] : never) : never;
export type ICharactersList<T> = T extends ICharacterBase<infer _S> ? T[] : never;

// https://github.com/Microsoft/TypeScript/issues/5579
export interface ICollocationState<T extends string = string> {
  readonly left: T[];
  readonly right: T[];
  readonly boat: T[];
  readonly boatPosition: RIVERSIDE_LEFT_TYPE | RIVERSIDE_RIGHT_TYPE;
  readonly isBoatInvalid: boolean;
}
