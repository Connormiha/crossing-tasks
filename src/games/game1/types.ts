import {ICharactersList, ICharacterBase} from 'flux/types';

export type ICharacterIdGame1 = 'farmer' | 'wolf' | 'cabbage' | 'sheep';
export interface ICharacterGame1 extends ICharacterBase<ICharacterIdGame1> {
    name: string;
}

export type ICharactersListGame1 = ICharactersList<ICharacterGame1>;
