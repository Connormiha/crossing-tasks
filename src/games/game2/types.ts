import {ICharactersList, ICharacterBase} from 'flux/types';

export type ICharacterIdGame2 = 'caveman_1' | 'caveman_2' | 'caveman_3' | 'priest_1' | 'priest_2' | 'priest_3';
export interface ICharacterGame2 extends ICharacterBase<ICharacterIdGame2> {
    name: string;
    type: string;
};
export type ICharactersListGame2 = ICharactersList<ICharacterGame2>;
