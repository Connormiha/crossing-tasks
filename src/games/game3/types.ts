import {ICharactersList, ICharacterBase} from 'flux/types';

export type ICharacterIdGame3 = 'women_red' | 'men_red' | 'women_white' | 'men_white' | 'women_black' | 'men_black';
export interface ICharacterGame3 extends ICharacterBase<ICharacterIdGame3> {
    name: string;
    family: 'red' | 'white' | 'black';
    sex: 'male' | 'female';
};

export type ICharactersListGame3 = ICharactersList<ICharacterGame3>;
