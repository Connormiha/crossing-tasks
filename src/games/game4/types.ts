import {ICharactersList, ICharacterBase} from 'flux/types';

export type ICharacterIdGame4 = 'men_black' | 'boy_red' | 'boy_yellow' | 'woman_black' | 'girl_red' | 'girl_yellow' | 'criminal' | 'policeman';
export interface ICharacterGame4 extends ICharacterBase<ICharacterIdGame4> {
    name: string;
    adult: boolean;
};

export type ICharactersListGame3 = ICharactersList<ICharacterGame4>;
