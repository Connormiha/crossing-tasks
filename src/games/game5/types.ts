import {ICharactersList, ICharacterBase} from 'flux/types';

export type ICharacterIdGame5 = 'priest_1' | 'priest_2' | 'priest_3' | 'monkey_1' | 'monkey_2' | 'gorilla';
export interface ICharacterGame5 extends ICharacterBase<ICharacterIdGame5> {
    name: string;
    type: 'monkey' | 'priest';
};

export type ICharactersListGame5 = ICharactersList<ICharacterGame5>;
