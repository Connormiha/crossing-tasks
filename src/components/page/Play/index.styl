.page
    padding: 20px

.content
    display: flex
    position: relative
    max-width: 700px
    min-height: 600px
    margin: 0 auto
    background: #4a8840 url('/src/assets/images/river.png') no-repeat
