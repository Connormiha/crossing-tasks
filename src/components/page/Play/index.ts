import { connect } from 'react-redux';
import { batchActions } from 'redux-batched-actions';
import * as collocationActions from 'flux/collocation';
import * as messageActions from 'flux/message';
import * as gameActions from 'flux/game';
import * as settingsActions from 'flux/settings';
import PagePlayPure from './index.pure';
import { IGameState } from 'flux/types';

import games from 'games';

export const mapDispatchToProps = (dispatch) => {
  return {
    onStartGame(id: string) {
      dispatch(
        batchActions([
          collocationActions.init(games[id].collocation),
          gameActions.set(id),
          collocationActions.toggleBoatInvalid(false),
        ]),
      );
    },

    onMoveCharacter(collocation: any, gameId: string, id: string) {
      const { success, message } = games[gameId].landingValidator(collocation, id);

      if (success) {
        dispatch(
          batchActions([
            collocationActions.moveCharacter(id),
            messageActions.set(''),
            collocationActions.toggleBoatInvalid(false),
          ]),
        );
      } else {
        dispatch(
          batchActions([messageActions.set(message), collocationActions.toggleBoatInvalid(true)]),
        );
      }
    },

    onMoveBoat(collocation: any, gameId: string) {
      const { success, message } = games[gameId].depetureValidator(collocation);

      if (success) {
        dispatch(
          batchActions([
            collocationActions.moveBoat(),
            messageActions.set(''),
            collocationActions.toggleBoatInvalid(false),
          ]),
        );
      } else {
        dispatch(
          batchActions([messageActions.set(message), collocationActions.toggleBoatInvalid(true)]),
        );
      }
    },

    onBoatMoveEnd(collocation: any) {
      const actions: any[] = [collocationActions.toggleBoatInvalid(false)];

      for (const id of collocation.boat) {
        actions.push(collocationActions.moveCharacter(id));
      }

      dispatch(batchActions(actions));
    },

    onToggleInvalidBoat(isBoatInvalid: boolean) {
      dispatch(collocationActions.toggleBoatInvalid(isBoatInvalid));
    },

    onFinishGame() {
      dispatch(gameActions.finish());
    },

    onChangeVolume(volume: number) {
      dispatch(settingsActions.setVolume(volume));
    },
  };
};

export default connect((state: IGameState) => state, mapDispatchToProps)(PagePlayPure);
