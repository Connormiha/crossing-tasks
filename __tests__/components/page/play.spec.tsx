import * as renderer from 'react-test-renderer';
import {mapDispatchToProps} from 'components/page/Play';
import PagePlayPure from 'components/page/Play/index.pure';
import {createAppStore} from 'store';
import noop from 'lodash/noop';
import games from 'games';
import {RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT} from 'flux/types';

/* tslint:disable:jsx-wrap-multiline */
describe('<PagePlayPure />', () => {
    const params = {
        game: {
            finished: false,
            currentGame: 'game_1'
        },
        message: {
            content: ''
        }
        match: {
            params: {
                id: ''
            }
        },
        settings: {
            volume: '1',
        },
        collocation: {
            [BOAT]: ['farmer', 'cabbage'],
            [RIVERSIDE_LEFT]: ['sheep'],
            [RIVERSIDE_RIGHT]: ['wolf'],
            boatPosition: RIVERSIDE_LEFT
        },
        onMoveCharacter: noop,
        onBoatMoveEnd: noop,
        onFinishGame: noop,
        onStartGame: noop,
        onMoveBoat: noop
    };

    it('should render', () => {
        let tree = renderer.create(
            <PagePlayPure
                {...params}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with message', () => {
        let tree = renderer.create(
            <PagePlayPure
                {...params}
                message={{content: 'Test'}}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render (Boat right)', () => {
        let tree = renderer.create(
            <PagePlayPure
                {...params}
                collocation={{...params.collocation, boatPosition: RIVERSIDE_RIGHT}}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render finished', () => {
        let tree = renderer.create(
            <PagePlayPure
                {...params}
                game={{...params.game, finished: true}}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});

describe('<PagePlay />', () => {
    let store,
        mapProps;

    const getState = () =>
        store.getState();

    beforeEach(() => {
        store = createAppStore();
        mapProps = mapDispatchToProps(store.dispatch);
    });

    it('should work onStartGame', () => {
        let game_1: string = Object.keys(games)[0];

        mapProps.onStartGame(game_1);

        expect(getState().collocation).toEqual(
            {...games[game_1].collocation, boatPosition: RIVERSIDE_LEFT}
        );

        expect(getState().game).toMatchObject({currentGame: game_1, finished: false});
    });
});
