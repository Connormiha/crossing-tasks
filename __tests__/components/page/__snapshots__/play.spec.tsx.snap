// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PagePlayPure /> should render (Boat right) 1`] = `
<div
  className="page-play"
>
  <div
    className="page-play__content"
  >
    <div
      className="riverside riverside_side_left"
    >
      <div>
        <div
          className="character character_skin_sheep"
          onClick={[Function]}
          title="Sheep"
        />
      </div>
    </div>
    <div
      className="riverside riverside_side_right"
    >
      <div>
        <div
          className="character character_skin_wolf"
          onClick={[Function]}
          title="Wolf"
        />
      </div>
    </div>
    <div
      className="boat boat_position_right"
      onTransitionEnd={[Function]}
    >
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_farmer character_packed"
          onClick={[Function]}
          title="Farmer"
        />
      </div>
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_cabbage character_packed"
          onClick={[Function]}
          title="Cabbage"
        />
      </div>
    </div>
    <button
      className="remote"
      disabled={false}
      onClick={[Function]}
    >
      Go
    </button>
    
  </div>
</div>
`;

exports[`<PagePlayPure /> should render 1`] = `
<div
  className="page-play"
>
  <div
    className="page-play__content"
  >
    <div
      className="riverside riverside_side_left"
    >
      <div>
        <div
          className="character character_skin_sheep"
          onClick={[Function]}
          title="Sheep"
        />
      </div>
    </div>
    <div
      className="riverside riverside_side_right"
    >
      <div>
        <div
          className="character character_skin_wolf"
          onClick={[Function]}
          title="Wolf"
        />
      </div>
    </div>
    <div
      className="boat boat_position_left"
      onTransitionEnd={[Function]}
    >
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_farmer character_packed"
          onClick={[Function]}
          title="Farmer"
        />
      </div>
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_cabbage character_packed"
          onClick={[Function]}
          title="Cabbage"
        />
      </div>
    </div>
    <button
      className="remote"
      disabled={false}
      onClick={[Function]}
    >
      Go
    </button>
    
  </div>
</div>
`;

exports[`<PagePlayPure /> should render finished 1`] = `
<div
  className="page-play"
>
  <div
    className="page-play__content"
  >
    <div
      className="riverside riverside_side_left"
    >
      <div>
        <div
          className="character character_skin_sheep"
          onClick={[Function]}
          title="Sheep"
        />
      </div>
    </div>
    <div
      className="riverside riverside_side_right"
    >
      <div>
        <div
          className="character character_skin_wolf"
          onClick={[Function]}
          title="Wolf"
        />
      </div>
    </div>
    <div
      className="boat boat_position_left"
      onTransitionEnd={[Function]}
    >
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_farmer character_packed"
          onClick={[Function]}
          title="Farmer"
        />
      </div>
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_cabbage character_packed"
          onClick={[Function]}
          title="Cabbage"
        />
      </div>
    </div>
    <button
      className="remote"
      disabled={false}
      onClick={[Function]}
    >
      Go
    </button>
    
    <div
      className="page-play__finish"
    >
      <div>
        Finished!
      </div>
    </div>
  </div>
</div>
`;

exports[`<PagePlayPure /> should render with message 1`] = `
<div
  className="page-play"
>
  <div
    className="page-play__content"
  >
    <div
      className="riverside riverside_side_left"
    >
      <div>
        <div
          className="character character_skin_sheep"
          onClick={[Function]}
          title="Sheep"
        />
      </div>
    </div>
    <div
      className="riverside riverside_side_right"
    >
      <div>
        <div
          className="character character_skin_wolf"
          onClick={[Function]}
          title="Wolf"
        />
      </div>
    </div>
    <div
      className="boat boat_position_left boat_invalid"
      onTransitionEnd={[Function]}
    >
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_farmer character_packed"
          onClick={[Function]}
          title="Farmer"
        />
      </div>
      <div
        className="boat__character"
      >
        <div
          className="character character_skin_cabbage character_packed"
          onClick={[Function]}
          title="Cabbage"
        />
      </div>
    </div>
    <button
      className="remote"
      disabled={false}
      onClick={[Function]}
    >
      Go
    </button>
    <div
      className="warning"
    >
      Test
    </div>
  </div>
</div>
`;
